File: solutions/ts/17.ts
function hasFourLights(board: string[][]): boolean {
  const horizontalLength = board[0].length;
  for (const row of board) {
    const line = row.join("");
    if (line.includes("RRRR") || line.includes("GGGG")) {
      return true;
    }
  }
  for (let i = 0; i < horizontalLength; i++) {
    let prev = "";
    let count = 0;
    for (const row of board) {
      const cell = row[i];
      if ((cell === "R" || cell === "G") && cell === prev) {
        count += 1;
        if (count === 4) return true;
      } else {
        prev = cell;
        count = cell === "R" || cell === "G" ? 1 : 0;
      }
    }
  }
  return false;
}

const has1 = hasFourLights([
  [".", ".", ".", ".", "."],
  ["R", "R", "R", "R", "."],
  ["G", "G", ".", ".", "."],
]);
// true → there are 4 red lights horizontally
console.log(has1);

const has2 = hasFourLights([
  [".", "G", ".", "."],
  [".", "G", ".", "."],
  [".", "G", ".", "."],
  [".", "G", ".", "."],
]);
// true → there are 4 green lights vertically
console.log(has2);

const has3 = hasFourLights([
  ["R", "G", "R"],
  ["G", "R", "G"],
  ["G", "R", "G"],
]);
// false → there are no 4 lights of the same color in a row
console.log(has3);

/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 150
  ops: 2983
*/

File: solutions/python/17.py
from functools import reduce


def has_four_lights(board: list[list[str]]) -> bool:
    for row in board:
        line = "".join(row)
        if "RRRR" in line or "GGGG" in line:
            return True
    horizontal_length = len(board[0])
    for i in range(0, horizontal_length):
        line = reduce(lambda curr, row: curr + row[i], board, "")
        if "RRRR" in line or "GGGG" in line:
            return True
    return False


has1 = has_four_lights([
    [".", ".", ".", ".", "."],
    ["R", "R", "R", "R", "."],
    ["G", "G", ".", ".", "."],
])
# true → there are 4 red lights horizontally
print(has1)

has2 = has_four_lights([
    [".", "G", ".", "."],
    [".", "G", ".", "."],
    [".", "G", ".", "."],
    [".", "G", ".", "."],
])
# true → there are 4 green lights vertically
print(has2)

has3 = has_four_lights([
    ["R", "G", "R"],
    ["G", "R", "G"],
    ["G", "R", "G"],
])
# false → there are no 4 lights of the same color in a row
print(has3)


"""
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 300
  ops: 3888
"""
